File: 2019/qualify-round/nodejs/src/slide-show.js
"use strict";

module.exports = class SlideShow {

  constructor(slideList) {
    this.slideList = slideList || [];
  }

  addSlide(slide) {
    this.slideList.push(slide);
  }

  getBestSlideCombination(elementIndex, array) {
    let index = -1, maximun = -Infinity;
    const element = array[elementIndex];
    for (let i = 0; i < array.length; i++) {
      if (!array[i].taken) {
        const tagsOne = this.getExludedTags(element, array[i]);
        const sameTags = element.mergedTags.size - tagsOne;
        const tagsTwo = array[i].mergedTags.size - sameTags;
        const minimun = Math.min(tagsOne, Math.min(tagsTwo, sameTags));
        if (minimun > maximun) {
          maximun = minimun;
          index = i;
        }
        console.log(elementIndex,maximun);
      }
    }
    return index;
  }

  getExludedTags(slideOne, slideTwo) {
    let counter = 0;
    for (let [key, value] of slideOne.mergedTags) {
      counter += !slideTwo.mergedTags.has(key) ? 1 : 0;
    }
    return counter;
  }

  generateFinalSlideShow() {
    const finalSlideShow = [];
    const i = this.getNextSlide();
    this.slideList[i].taken = true;
    finalSlideShow.push(this.slideList[i]);
    while (i >= 0) {
      const index = this.getBestSlideCombination(i, this.slideList);
      if (index >= 0) {
        this.slideList[index].taken = true;
        finalSlideShow.push(this.slideList[index]);
      }
      i=index;
    }

    return finalSlideShow;
  }

  getNextSlide() {
    let i = 0, found = false;
    while (!found && i < this.slideList.length) {
      found = !this.slideList[i].taken;
      i++;
    }
    return found ? i - 1 : -1;
  }

}

File: 2019/qualify-round/java/src/hashCode/SlideShow.java
package hashCode;

import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

public class SlideShow {

	private List<Slide> slideList;

	public List<Slide> getSlideList() {
		return slideList;
	}

	public SlideShow() {
		this.slideList = new LinkedList<Slide>();
	}

	public void addSlide(Slide slide) {
		this.slideList.add(slide);
	}

	public LinkedList<Slide> generateFinalSlideShow() {
		LinkedList<Slide> finalSlideShow = new LinkedList<Slide>();
		int count = 0;
		Slide slide = this.getNextSlide();
		slide.setTaken(true);
		finalSlideShow.add(slide);
		while (slide!=null) {
			System.out.println(count);
			count++;
			Slide slideCombination = this.getBestSlideCombination(slide);
			if (slideCombination != null) {
				slideCombination.setTaken(true);
				finalSlideShow.add(slideCombination);
			} 
			slide = slideCombination;
		}

		return finalSlideShow;
	}

	@Override
	public String toString() {
		return "SlideShow [slideList=" + slideList + "]";
	}

	private Slide getBestSlideCombination(Slide slide) {
		Slide foundSlide = null;
		int max = Integer.MIN_VALUE;
		Iterator<Slide> it = this.slideList.iterator();
		while (it.hasNext()) {
			Slide current = it.next();
			if (!current.isTaken()) {
				int tagsOne = this.getExludedTags(slide, current);
				int sameTags = slide.getMergedTags().size() - tagsOne;
				int tagsTwo = current.getMergedTags().size() - sameTags;
				int min = Integer.min(tagsOne, Integer.min(sameTags, tagsTwo));
				if (min > max) {
					max = min;
					foundSlide = current;
				}
			}
		}

		return foundSlide;

	}

	private int getExludedTags(Slide slideOne, Slide slideTwo) {
		Set<String> diff = new HashSet<String>(slideOne.getMergedTags());
		diff.removeAll(slideTwo.getMergedTags());
		return diff.size();
	}

	private Slide getNextSlide() {
		boolean found = false;
		Slide slide = null;
		Iterator<Slide> it = this.slideList.iterator();
		while (!found && it.hasNext()) {
			slide = it.next();
			found = !slide.isTaken();
		}
		return found ? slide : null;
	}

}
